src/transformers/models/whisper/modeling_whisper.py

```diff
@@ -1364,25 +1360,27 @@ def forward(
             raise ValueError("You have to specify either decoder_input_ids or decoder_inputs_embeds")
 
         # past_key_values_length
-        past_key_values_length = past_key_values[0][0].shape[2] if past_key_values is not None else 0
+        past_key_values_length = 0
+        if use_cache:
+            use_legacy_cache = not (past_key_values is not None and isinstance(past_key_values[0], Cache))
+            if use_legacy_cache:
+                if past_key_values is None:
+                    self_attn = cross_attn = None
+                else:
+                    self_attn = [key_values[:2] for key_values in past_key_values]
+                    cross_attn = [key_values[2:] for key_values in past_key_values]
+                past_key_values = (
+                    DynamicCache.from_legacy_cache(self_attn),
+                    DynamicCache.from_legacy_cache(cross_attn),
+                )
+            past_key_values_length = past_key_values[0].get_seq_length()
```

Comment:
we should abstract a little bit this function because other encoder-decoder models will need it